knowrob_srdl/src/urdf_parser.cpp

```diff
@@ -0,0 +1,852 @@
+        return true;
+    } catch (const std::runtime_error& e) {
+        ROS_ERROR("%s", e.what());
+        return false;
+    }
+}
+
+PREDICATE(link_material_color, 3) {
+    try {
+        std::string link_name((char*) PL_A1);
+        long index = (long) PL_A2;
+        urdf::LinkConstSharedPtr link = get_link(link_name);
+        if (!link_has_visual_with_index(link, index) ||
+                !link->visual_array[index]->material)
+            return false;
+        PL_A3 = to_prolog_rgba(link->visual_array[index]->material->color);
+        return true;
+    } catch (const std::runtime_error& e) {
+        ROS_ERROR("%s", e.what());
+        return false;
+    }
+}
+
+PREDICATE(link_material_texture, 3) {
+    try {
+        std::string link_name((char*) PL_A1);
+        long index = (long) PL_A2;
+        urdf::LinkConstSharedPtr link = get_link(link_name);
+        if (!link_has_visual_with_index(link, index) ||
+                !link->visual_array[index]->material ||
+                link->visual_array[index]->material->texture_filename.compare("") == 0)
+            return false;
+        PL_A3 = link->visual_array[index]->material->texture_filename.c_str();
+        return true;
+    } catch (const std::runtime_error& e) {
+        ROS_ERROR("%s", e.what());
+        return false;
+    }
+}
+
+PREDICATE(link_num_collisions, 2) {
+    try {
+        std::string link_name((char*) PL_A1);
+        urdf::LinkConstSharedPtr link = get_link(link_name);
+        PL_A2 = (long) link->collision_array.size();
+        return true;
+    } catch (const std::runtime_error& e) {
+        ROS_ERROR("%s", e.what());
+        return false;
+    }
+}
+
+PREDICATE(link_collision_type, 3) {
+    try {
+        std::string link_name((char*) PL_A1);
+        long index = (long) PL_A2;
+        urdf::LinkConstSharedPtr link = get_link(link_name);
+        if (!link_has_collision_with_index(link, index) ||
+                !link->collision_array[index]->geometry)
+            return false;
+        switch (link->collision_array[index]->geometry->type) {
+            case urdf::Geometry::BOX: {
+                PL_A3 = "box";
+                return true;
+            }
+            case urdf::Geometry::CYLINDER: {
+                PL_A3 = "cylinder";
+                return true;
+            }
+            case urdf::Geometry::SPHERE: {
+                PL_A3 = "sphere";
+                return true;
+            }
+            case urdf::Geometry::MESH: {
+                PL_A3 = "mesh";
+                return true;
+            }
+            default:
+                return false;
+        }
+    } catch (const std::runtime_error& e) {
+        ROS_ERROR("%s", e.what());
+        return false;
+    }
+}
+
+PREDICATE(link_collision_name, 3) {
+    try {
+        std::string link_name((char*) PL_A1);
+        long index = (long) PL_A2;
+        urdf::LinkConstSharedPtr link = get_link(link_name);
```

Comment:
the shared ptr has some reference counting inside and auto deletes if counter reaches zero I guess?